### PR TITLE
add raster basics to datatypes

### PR DESCRIPTION
**Pull request by jdroenner**
raster concepts need more work

### CODE REVIEW

File: datatypes/src/raster/raster.rs [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+    ///    vec![1,2,3,4,5,6],
+    ///    None,
+    ///    TimeInterval::default(),
+    ///    [1.0, 1.0, 0.0, 1.0, 0.0, 1.0].into(),
+    /// ).unwrap();
+    /// raster2d.set_pixel_value_at_grid_index(&(1, 1), 9).unwrap();
+    /// assert_eq!(raster2d.data_container(), &[1,2,3,9,5,6]);
+    /// ```
+    ///
+    fn set_pixel_value_at_grid_index(&mut self, grid_index: &I, value: T) -> Result<()>;
```
Comment (ChristianBeilschmidt) [inline datatypes/src/raster/raster.rs]: Is immutability an important constraint for our types? If yes, we should do something builder like with a mutable grid.

File: datatypes/src/raster/raster.rs [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+        let raster2d = SimpleRaster2d::new(
+            dim.into(),
+            data,
+            None,
+            temporal_bounds,
+            geo_transform.into(),
+        )
+        .unwrap();
+        let value = raster2d.pixel_value_at_grid_index(&tuple_index).unwrap();
+        assert!(value == 4);
```
Comment (ChristianBeilschmidt) [inline datatypes/src/raster/raster.rs]: Just as a general suggestion, why don't you just use
```suggestion
        assert_eq!(
            raster2d.pixel_value_at_grid_index(&tuple_index).unwrap(),
            4
        );
```

File: datatypes/src/primitives/bounding_box.rs [hunk 3]
```diff
@@ -0,0 +1,416 @@
...
+    /// let bbox = BoundingBox2D::new(ll, ur).unwrap();
+    /// ```
+    ///
+    pub fn new(
+        lower_left_coordinate: Coordinate2D,
+        upper_right_coordinate: Coordinate2D,
+    ) -> Result<Self> {
+        ensure!(
+            lower_left_coordinate.x <= upper_right_coordinate.x
+                && lower_left_coordinate.y <= upper_right_coordinate.x,
```
[the remark this answers is not in the document]
Comment (jdroenner) [reply datatypes/src/primitives/bounding_box.rs]: yip

File: datatypes/src/primitives/bounding_box.rs [hunk 4]
```diff
@@ -0,0 +1,416 @@
...
+    ///
+    /// let ll = Coordinate2D::new(1.0, 1.0);
+    /// let ur = Coordinate2D::new(4.0, 4.0);
+    /// let bbox = BoundingBox2D::new(ll, ur).unwrap();
+    ///
+    /// let ll_in = Coordinate2D::new(2.0, 2.0);
+    /// let ur_in = Coordinate2D::new(3.0, 3.0);
+    /// let bbox_in = BoundingBox2D::new(ll, ur).unwrap();
+    ///
+    /// assert!(bbox.contains_bbox(&bbox_in));
```
[the remark this answers is not in the document]
Comment (jdroenner) [reply datatypes/src/primitives/bounding_box.rs]: done

File: datatypes/src/primitives/bounding_box.rs [hunk 5]
```diff
@@ -0,0 +1,416 @@
...
+    /// use geoengine_datatypes::primitives::{Coordinate2D, BoundingBox2D};
+    ///
+    /// let ll = Coordinate2D::new(1.0, 1.0);
+    /// let ur = Coordinate2D::new(4.0, 4.0);
+    /// let bbox = BoundingBox2D::new(ll, ur).unwrap();
+    ///
+    /// let ll_inside = Coordinate2D::new(2.0, 2.0);
+    /// let ur_inside = Coordinate2D::new(3.0, 3.0);
+    /// let bbox_inside = BoundingBox2D::new(ll_inside, ur_inside).unwrap();
+    /// assert!(bbox.overlaps_bbox(&bbox_inside));
```
[the remark this answers is not in the document]
Comment (jdroenner) [reply datatypes/src/primitives/bounding_box.rs]: done

File: datatypes/src/raster/geo_transform.rs [hunk 6]
```diff
@@ -0,0 +1,110 @@
+use crate::primitives::Coordinate2D;
+
+/// This is a typedef for the GDAL GeoTransform. It represents an affine transformation matrix.
+pub type GdalGeoTransform = [f64; 6];
+
+/// The GeoTransform is a more user friendly representation of the GDAL GeoTransform affine transformation matrix.
+#[derive(Copy, Clone, PartialEq, Debug)]
+pub struct GeoTransform {
+    upper_left_x_coordinate: f64,
```
[the remark this answers is not in the document]
Comment (jdroenner) [reply datatypes/src/raster/geo_transform.rs]: done and added additional censtructor

File: datatypes/src/raster/geo_transform.rs [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+    /// use geoengine_datatypes::raster::{GeoTransform};
+    /// use geoengine_datatypes::primitives::{Coordinate2D};
+    ///
+    /// let geo_transform = GeoTransform::new(0.0, 1.0, 0.0, -1.0);
+    /// assert_eq!(geo_transform.grid_2d_to_coordinate_2d((0, 0)), (0.0, 0.0).into())
+    /// ```
+    ///
+    pub fn grid_2d_to_coordinate_2d(&self, grid_index: (usize, usize)) -> Coordinate2D {
+        let (grid_index_y, grid_index_x) = grid_index;
+        let coord_x = self.upper_left_x_coordinate + (grid_index_x as f64) * self.x_pixel_size; //  + (grid_index_y as f64) * self.x_rotation;
```
[the remark this answers is not in the document]
Comment (jdroenner) [reply datatypes/src/raster/geo_transform.rs]: changed this to point to the gdal docs.

File: datatypes/src/raster/raster.rs (hunk 1 above)
Comment (jdroenner) [reply datatypes/src/raster/raster.rs]: i guess we have to discuss this

**Pull request merged**